crates/sui-core/src/transaction_manager.rs

```diff
@@ -108,7 +112,116 @@ impl Inner {
+                .is_none());
+            // When a certificate has all locks acquired, it is ready to execute.
+            if pending_cert.acquiring_locks.is_empty() {
+                let pending_cert = self.pending_certificates.remove(&digest).unwrap();
+                ready_certificates.push(pending_cert);
+            } else {
+                // TODO: we should start logging this at a higher level after some period of
+                // time has elapsed.
+                debug!(tx_digest = ?digest,acquiring = ?pending_cert.acquiring_locks, "Certificate acquiring locks");
+            }
+        }
+
+        ready_certificates
+    }
+
+    /// After reaching 3/4 load in hashmaps, increase capacity to decrease load to 1/2.
+    fn maybe_reserve_capacity(&mut self) {
+        if self.lock_waiters.len() > self.lock_waiters.capacity() * 3 / 4 {
+            self.lock_waiters.reserve(self.lock_waiters.capacity() / 2);
+        }
+        if self.input_objects.len() > self.input_objects.capacity() * 3 / 4 {
+            self.input_objects
+                .reserve(self.input_objects.capacity() / 2);
+        }
+        if self.pending_certificates.len() > self.pending_certificates.capacity() * 3 / 4 {
+            self.pending_certificates
+                .reserve(self.pending_certificates.capacity() / 2);
+        }
+        if self.executing_certificates.len() > self.executing_certificates.capacity() * 3 / 4 {
+            self.executing_certificates
+                .reserve(self.executing_certificates.capacity() / 2);
+        }
+    }
+
+    /// After reaching 1/4 load in hashmaps, decrease capacity to increase load to 1/2.
+    fn maybe_shrink_capacity(&mut self) {
+        if self.lock_waiters.len() > MIN_HASHMAP_CAPACITY
+            && self.lock_waiters.len() < self.lock_waiters.capacity() / 4
+        {
+            self.lock_waiters
+                .shrink_to(max(self.lock_waiters.capacity() / 2, MIN_HASHMAP_CAPACITY))
+        }
+        if self.input_objects.len() > MIN_HASHMAP_CAPACITY
+            && self.input_objects.len() < self.input_objects.capacity() / 4
+        {
+            self.input_objects
+                .shrink_to(max(self.input_objects.capacity() / 2, MIN_HASHMAP_CAPACITY))
+        }
+        if self.pending_certificates.len() > MIN_HASHMAP_CAPACITY
+            && self.pending_certificates.len() < self.pending_certificates.capacity() / 4
+        {
+            self.pending_certificates.shrink_to(max(
+                self.pending_certificates.capacity() / 2,
+                MIN_HASHMAP_CAPACITY,
+            ))
+        }
+        if self.executing_certificates.len() > MIN_HASHMAP_CAPACITY
+            && self.executing_certificates.len() < self.executing_certificates.capacity() / 4
+        {
+            self.executing_certificates.shrink_to(max(
+                self.executing_certificates.capacity() / 2,
+                MIN_HASHMAP_CAPACITY,
+            ))
```

Comment:
Can you explain more about what memory problem this is addressing? It seems to me like the main problem would be peak usage, which this doesn't seem to help with.

Also, this code is quite repetive and could be factored with a generic function.

Reply:
This tries to reduce memory usage after a spike of pending transactions, which can be much larger than the usual number of pending transactions. Will refactor this.